modules/network/src/main/java/org/apache/ignite/internal/network/serialization/ClassDescriptorParser.java

```diff
@@ -0,0 +1,432 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.network.serialization;
+
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.ObjectStreamException;
+import java.io.Serializable;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import org.apache.ignite.lang.IgniteException;
+
+/**
+ * Class parser for the user object serialization.
+ */
+public class ClassDescriptorParser {
+
+    /**
+     * Parser context.
+     */
+    private final ClassDescriptorParserContext context;
+
+    /**
+     * Constructor.
+     *
+     * @param ctx Context.
+     */
+    public ClassDescriptorParser(ClassDescriptorParserContext ctx) {
+        this.context = ctx;
+    }
+
+    /**
+     * Parses the class definition and definitions of classes fields if they're not already parsed.
+     *
+     * @param clazz Class definition.
+     * @return Class descriptor.
+     */
+    public ClassDescriptor parse(Class<?> clazz) {
+        ClassDescriptor classDesc = parse0(clazz);
+
+        Map<Integer, ClassDescriptor> localDescs = new HashMap<>();
+        localDescs.put(classDesc.descriptorId(), classDesc);
+
+        Queue<FieldDescriptor> fieldDescriptors = new LinkedList<>(classDesc.fields());
+
+        while (!fieldDescriptors.isEmpty()) {
+            FieldDescriptor fieldDescriptor = fieldDescriptors.remove();
+
+            int typeDescriptorId = fieldDescriptor.typeDescriptorId();
+
+            if (context.hasDescriptor(typeDescriptorId) || localDescs.containsKey(
+                    typeDescriptorId)) {
+                continue;
+            }
+
+            Class<?> fieldClass = fieldDescriptor.clazz();
+
+            ClassDescriptor fieldClassDesc = parse0(fieldClass);
+
+            localDescs.put(typeDescriptorId, fieldClassDesc);
+
+            fieldDescriptors.addAll(fieldClassDesc.fields());
+        }
+
+        context.addDescriptors(localDescs.values());
+
+        return classDesc;
+    }
+
+    /**
+     * Parses the class definition.
+     *
+     * @param clazz Class.
+     * @return Class descriptor.
+     */
+    private ClassDescriptor parse0(Class<?> clazz) {
+        assert !clazz.isPrimitive() :
```

Comment:
Well, class descriptors for primitives should be already registered, so this should never happen. A test should be covering this situation (and failing with the error). But of course it can be changed

Reply:
Ok